src/main/java/org/tikv/common/failsafe/CircuitBreakerImpl.java

```diff
@@ -0,0 +1,187 @@
+  private final int sleepWindowInSeconds;
+  private final int attemptRequestCount;
+
+  private final AtomicLong circuitOpened = new AtomicLong(-1);
+  private final AtomicReference<Status> status = new AtomicReference<>(Status.CLOSED);
+  private final AtomicLong attemptCount = new AtomicLong(0);
+  private final AtomicLong attemptSuccessCount = new AtomicLong(0);
+
+  private final CircuitBreakerMetrics metrics;
+
+  public CircuitBreakerImpl(TiConfiguration conf) {
+    this(
+        conf.isCircuitBreakEnable(),
+        conf.getCircuitBreakAvailabilityWindowInSeconds(),
+        conf.getCircuitBreakAvailabilityErrorThresholdPercentage(),
+        conf.getCircuitBreakAvailabilityRequestVolumnThreshold(),
+        conf.getCircuitBreakSleepWindowInSeconds(),
+        conf.getCircuitBreakAttemptRequestCount());
+  }
+
+  public CircuitBreakerImpl(
+      boolean enable,
+      int windowInSeconds,
+      int errorThresholdPercentage,
+      int requestVolumeThreshold,
+      int sleepWindowInSeconds,
+      int attemptRequestCount) {
+    this.enable = enable;
+    this.windowInSeconds = windowInSeconds;
+    this.errorThresholdPercentage = errorThresholdPercentage;
+    this.requestVolumeThreshold = requestVolumeThreshold;
+    this.sleepWindowInSeconds = sleepWindowInSeconds;
+    this.attemptRequestCount = attemptRequestCount;
+    this.metrics = new CircuitBreakerMetricsImpl(windowInSeconds);
+    this.metrics.addListener(getMetricsListener());
+  }
+
+  private MetricsListener getMetricsListener() {
+    return hc -> {
+      logger.info("onNext " + hc.toString());
+      // check if we are past the requestVolumeThreshold
+      if (hc.getTotalRequests() < requestVolumeThreshold) {
+        // we are not past the minimum volume threshold for the stat window,
+        // so no change to circuit status.
+        // if it was CLOSED, it stays CLOSED
+        // if it was half-open, we need to wait for some successful command executions
+        // if it was open, we need to wait for sleep window to elapse
+      } else {
+        if (hc.getErrorPercentage() < errorThresholdPercentage) {
+          // we are not past the minimum error threshold for the stat window,
+          // so no change to circuit status.
+          // if it was CLOSED, it stays CLOSED
+          // if it was half-open, we need to wait for some successful command executions
+          // if it was open, we need to wait for sleep window to elapse
+        } else {
+          // our failure rate is too high, we need to set the state to OPEN
+          close2Open();
+        }
+      }
+    };
+  }
+
+  @Override
+  public CircuitBreakerMetrics getMetrics() {
+    return metrics;
+  }
+
+  @Override
+  public boolean allowRequest() {
+    if (!enable) {
+      return true;
+    }
+    return !isOpen();
+  }
+
+  @Override
+  public boolean isOpen() {
+    return circuitOpened.get() >= 0;
+  }
+
+  @Override
+  public Status getStatus() {
```

Comment:
seems this function is only used by the test, can we remove it from the interface definition and make it visible only to test?

Reply:
sure